File: Modules/ServerTypes/Shift/ShiftForm.ts
import { IntegerEditor, StringEditor, TextAreaEditor, PrefixedContext } from "@serenity-is/corelib";
import { initFormType } from "@serenity-is/corelib/q";

export interface ShiftForm {
    TypeOfShift: IntegerEditor;
    ShiftColor: StringEditor;
    ShiftName: StringEditor;
    Description: TextAreaEditor;
    NightShiftBetweenStart: StringEditor;
    NightShiftBetweenEnd: StringEditor;
    SundayStartingFrom: StringEditor;
    SundayEndingAt: StringEditor;
    SundayWorkingTime: IntegerEditor;
    SundayLunchTimeStartingFrom: StringEditor;
    SundayLunchTimeEndingAt: StringEditor;
    MondayStartingFrom: StringEditor;
    MondayEndingAt: StringEditor;
    MondayWorkingTime: IntegerEditor;
    MondayLunchTimeStartingFrom: StringEditor;
    MondayLunchTimeEndingAt: StringEditor;
    TuesdayStartingFrom: StringEditor;
    TuesdayEndingAt: StringEditor;
    TuesdayWorkingTime: IntegerEditor;
    TuesdayLunchTimeStartingFrom: StringEditor;
    TuesdayLunchTimeEndingAt: StringEditor;
    WednesdayStartingFrom: StringEditor;
    WednesdayEndingAt: StringEditor;
    WednesdayWorkingTime: IntegerEditor;
    WednesdayLunchTimeStartingFrom: StringEditor;
    WednesdayLunchTimeEndingAt: StringEditor;
    ThursdayStartingFrom: StringEditor;
    ThursdayEndingAt: StringEditor;
    ThursdayWorkingTime: IntegerEditor;
    ThursdayLunchTimeStartingFrom: StringEditor;
    ThursdayLunchTimeEndingAt: StringEditor;
    FridayStartingFrom: StringEditor;
    FridayEndingAt: StringEditor;
    FridayWorkingTime: IntegerEditor;
    FridayLunchTimeStartingFrom: StringEditor;
    FridayLunchTimeEndingAt: StringEditor;
    SaturdayStartingFrom: StringEditor;
    SaturdayEndingAt: StringEditor;
    SaturdayWorkingTime: IntegerEditor;
    SaturdayLunchTimeStartingFrom: StringEditor;
    SaturdayLunchTimeEndingAt: StringEditor;
}

export class ShiftForm extends PrefixedContext {
    static formKey = 'Shift.Shift';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!ShiftForm.init)  {
            ShiftForm.init = true;

            var w0 = IntegerEditor;
            var w1 = StringEditor;
            var w2 = TextAreaEditor;

            initFormType(ShiftForm, [
                'TypeOfShift', w0,
                'ShiftColor', w1,
                'ShiftName', w1,
                'Description', w2,
                'NightShiftBetweenStart', w1,
                'NightShiftBetweenEnd', w1,
                'SundayStartingFrom', w1,
                'SundayEndingAt', w1,
                'SundayWorkingTime', w0,
                'SundayLunchTimeStartingFrom', w1,
                'SundayLunchTimeEndingAt', w1,
                'MondayStartingFrom', w1,
                'MondayEndingAt', w1,
                'MondayWorkingTime', w0,
                'MondayLunchTimeStartingFrom', w1,
                'MondayLunchTimeEndingAt', w1,
                'TuesdayStartingFrom', w1,
                'TuesdayEndingAt', w1,
                'TuesdayWorkingTime', w0,
                'TuesdayLunchTimeStartingFrom', w1,
                'TuesdayLunchTimeEndingAt', w1,
                'WednesdayStartingFrom', w1,
                'WednesdayEndingAt', w1,
                'WednesdayWorkingTime', w0,
                'WednesdayLunchTimeStartingFrom', w1,
                'WednesdayLunchTimeEndingAt', w1,
                'ThursdayStartingFrom', w1,
                'ThursdayEndingAt', w1,
                'ThursdayWorkingTime', w0,
                'ThursdayLunchTimeStartingFrom', w1,
                'ThursdayLunchTimeEndingAt', w1,
                'FridayStartingFrom', w1,
                'FridayEndingAt', w1,
                'FridayWorkingTime', w0,
                'FridayLunchTimeStartingFrom', w1,
                'FridayLunchTimeEndingAt', w1,
                'SaturdayStartingFrom', w1,
                'SaturdayEndingAt', w1,
                'SaturdayWorkingTime', w0,
                'SaturdayLunchTimeStartingFrom', w1,
                'SaturdayLunchTimeEndingAt', w1
            ]);
        }
    }
}

File: Modules/HumanResource/Shift/Shift/ShiftForm.cs
using Serenity.ComponentModel;
using System;
using System.ComponentModel;

namespace HRMSoftware.Shift.Forms;

[FormScript("Shift.Shift")]
[BasedOnRow(typeof(ShiftRow), CheckNames = true)]
public class ShiftForm
{
    public int TypeOfShift { get; set; }


    public string ShiftColor { get; set; }

    public string ShiftName { get; set; }
    [TextAreaEditor(Rows = 5)]
    public string Description { get; set; }

    [Category("Night Shift Categorisation")]
    public string NightShiftBetweenStart { get; set; }
    public string NightShiftBetweenEnd { get; set; }


    [Category("Sunday")]
    [HalfWidth]
    public string SundayStartingFrom { get; set; }
    [HalfWidth]
    public string SundayEndingAt { get; set; }
    [HalfWidth]
    public int SundayWorkingTime { get; set; }

    [HalfWidth]
    public string SundayLunchTimeStartingFrom { get; set; }
    [HalfWidth]
    public string SundayLunchTimeEndingAt { get; set; }

    [Category("Monday")]
    [HalfWidth]
    public string MondayStartingFrom { get; set; }
    [HalfWidth]
    public string MondayEndingAt { get; set; }

    [HalfWidth]
    public int MondayWorkingTime { get; set; }

    [HalfWidth]
    public string MondayLunchTimeStartingFrom { get; set; }
    [HalfWidth]
    public string MondayLunchTimeEndingAt { get; set; }


    [Category("Tuesday")]
    [HalfWidth]
    public string TuesdayStartingFrom { get; set; }
    [HalfWidth]
    public string TuesdayEndingAt { get; set; }
    [HalfWidth]
    public int TuesdayWorkingTime { get; set; }

    [HalfWidth]
    public string TuesdayLunchTimeStartingFrom { get; set; }
    [HalfWidth]
    public string TuesdayLunchTimeEndingAt { get; set; }


    [Category("Wednessday")]
    [HalfWidth]
    public string WednesdayStartingFrom { get; set; }
    [HalfWidth]
    public string WednesdayEndingAt { get; set; }
    [HalfWidth]
    public int WednesdayWorkingTime { get; set; }

    [HalfWidth]
    public string WednesdayLunchTimeStartingFrom { get; set; }
    [HalfWidth]
    public string WednesdayLunchTimeEndingAt { get; set; }


    [Category("Thursday")]
    [HalfWidth]
    public string ThursdayStartingFrom { get; set; }
    [HalfWidth]
    public string ThursdayEndingAt { get; set; }
    [HalfWidth]
    public int ThursdayWorkingTime { get; set; }

    [HalfWidth]
    public string ThursdayLunchTimeStartingFrom { get; set; }
    [HalfWidth]
    public string ThursdayLunchTimeEndingAt { get; set; }


    [Category("Friday")]
    [HalfWidth]
    public string FridayStartingFrom { get; set; }
    [HalfWidth]
    public string FridayEndingAt { get; set; }
    [HalfWidth]
    public int FridayWorkingTime { get; set; }

    [HalfWidth]
    public string FridayLunchTimeStartingFrom { get; set; }
    [HalfWidth]
    public string FridayLunchTimeEndingAt { get; set; }


    [Category("Saturday")]
    [HalfWidth]
    public string SaturdayStartingFrom { get; set; }
    [HalfWidth]
    public string SaturdayEndingAt { get; set; }
    [HalfWidth]
    public int SaturdayWorkingTime { get; set; }

    [HalfWidth]
    public string SaturdayLunchTimeStartingFrom { get; set; }
    [HalfWidth]
    public string SaturdayLunchTimeEndingAt { get; set; }






}